cmd/juju/application/removeapplication.go

```diff
@@ -59,48 +59,122 @@ func (c *removeApplicationCommand) Init(args []string) error {
+	DestroyUnitsDeprecated(unitNames ...string) error
+	GetCharmURL(appName string) (*charm.URL, error)
 	ModelUUID() string
 }
 
-func (c *removeApplicationCommand) getAPI() (removeApplicationAPI, error) {
+func (c *removeApplicationCommand) getAPI() (removeApplicationAPI, int, error) {
 	root, err := c.NewAPIRoot()
 	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, -1, errors.Trace(err)
 	}
-	return application.NewClient(root), nil
+	version := root.BestFacadeVersion("Application")
+	return application.NewClient(root), version, nil
 }
 
 func (c *removeApplicationCommand) Run(ctx *cmd.Context) error {
-	client, err := c.getAPI()
+	client, apiVersion, err := c.getAPI()
 	if err != nil {
 		return err
 	}
 	defer client.Close()
-	err = block.ProcessBlockedError(client.Destroy(c.ApplicationName), block.BlockRemove)
-	if err != nil {
-		return err
+
+	if apiVersion < 4 {
+		return c.removeApplicationsDeprecated(ctx, client)
 	}
-	return c.removeAllocation(ctx)
+	return c.removeApplications(ctx, client)
 }
 
-func (c *removeApplicationCommand) removeAllocation(ctx *cmd.Context) error {
-	client, err := c.getAPI()
-	if err != nil {
-		return err
+func (c *removeApplicationCommand) removeApplicationsDeprecated(
+	ctx *cmd.Context,
+	client removeApplicationAPI,
+) error {
+	for _, name := range c.ApplicationNames {
+		err := client.DestroyDeprecated(name)
+		if err := block.ProcessBlockedError(err, block.BlockRemove); err != nil {
+			return errors.Trace(err)
+		}
+		if err := c.removeAllocation(ctx, client, name); err != nil {
+			return errors.Trace(err)
+		}
 	}
-	charmURL, err := client.GetCharmURL(c.ApplicationName)
+	return nil
+}
+
+func (c *removeApplicationCommand) removeApplications(
+	ctx *cmd.Context,
+	client removeApplicationAPI,
+) error {
+	results, err := client.DestroyApplication(c.ApplicationNames...)
+	if err := block.ProcessBlockedError(err, block.BlockRemove); err != nil {
+		return errors.Trace(err)
+	}
+	anyFailed := false
+	for i, name := range c.ApplicationNames {
+		result := results[i]
+		if result.Error != nil {
+			ctx.Infof("removing application %s failed: %s", name, result.Error)
+			anyFailed = true
+			continue
+		}
+		if err := c.removeAllocation(ctx, client, name); err != nil {
+			ctx.Infof("removing allocation for application %s failed: %s", name, result.Error)
+			anyFailed = true
+			continue
+		}
+		ctx.Infof("removing application %s", name)
+		for _, entity := range result.Info.DestroyedUnits {
+			unitTag, err := names.ParseUnitTag(entity.Tag)
+			if err != nil {
+				logger.Warningf("%s", err)
+				continue
+			}
+			ctx.Verbosef("- removing %s", names.ReadableString(unitTag))
+		}
+		for _, entity := range result.Info.DestroyedStorage {
+			storageTag, err := names.ParseStorageTag(entity.Tag)
+			if err != nil {
+				logger.Warningf("%s", err)
+				continue
+			}
+			ctx.Infof("- removing %s", names.ReadableString(storageTag))
```

Comment:
Having it worded as "removing" will be confusing, because for destroy-controller, we say "destroying machine" or whatever and then when that all is done, the job is finished. But here that same expectation doesn't hold true.
I think "will remove blah" is better as it better communicates the async semantics of the operation.